browser/ui/webui/ai_chat/ai_chat_ui_page_handler_browsertest.cc

```diff
@@ -84,8 +84,14 @@ class AIChatUIPageHandlerBrowserTest : public PlatformBrowserTest,
   }
 
   mojom::TabDataPtr GetTabDataForFirstMatchingURL(const GURL& url) {
-    auto it = std::find_if(tabs_.begin(), tabs_.end(),
-                           [url](const auto& tab) { return tab->url == url; });
+    auto find_tab = [&]() {
+      return std::find_if(tabs_.begin(), tabs_.end(),
+                          [url](const auto& tab) { return tab->url == url; });
+    };
+
+    CHECK(base::test::RunUntil([&]() { return find_tab() != tabs_.end(); }));
```

Comment:
So why do we think this is not matching when this function is first called? web_contents() in the tests is waiting for load stop before the test is run. Is it the call to `SetNeedsReload` which makes the url not match? Or is it because `TabDataChanged` isn't called yet? Should we wait for that before starting each test?

Reply:
My theory is that `TabDataChanged` hasn't been called yet. `TabDataChanged` needs to be called for each added tab.

Reply:
One more question then - because this seems tied to how `TabTrackerService` works anyway, is it possible to have a `RunLoop` wait until `TabDataChanged` is called after the load and be more explicit about it?

Reply:
@petemill if possible, it'd be great to try and land it like this first, so we can see if that's the actual race.

Reply:
From my experience, mojo binding works in async. so I think `GetTabDataForFirstMatchingURL()` could be called before binding completes. I assume that waiting till `TabDataChanged()` is called before calling `GetTabDataForFirstMatchingURL()` could fix. and I think this waiting also has same effect as we only called once.